cmd/agent/app/throttling/throttler.go

```diff
@@ -0,0 +1,294 @@
+	var rateLimiter *tokenBucket
+	if a.perOperationRateLimiter.Has(operationName) ||
+		a.perOperationRateLimiter.IsFull() {
+		rateLimiter = a.perOperationRateLimiter.Get(operationName).(*tokenBucket)
+	} else {
+		rateLimiter = newTokenBucket(a.options, a.timeNow)
+		a.perOperationRateLimiter.Set(operationName, rateLimiter)
+	}
+	credits := rateLimiter.Withdraw(maxWithdrawal)
+	return credits
+}
+
+// ThrottlerOptions provides values to be used in a Throttler object.
+type ThrottlerOptions struct {
+	// DefaultAccountOptions defines the default AccountOptions to use for all
+	// service accounts.
+	DefaultAccountOptions AccountOptions
+	// AccountOptionOverrides overrides DefaultAccountOptions for services with
+	// service names present in the map.
+	AccountOptionOverrides map[string]*AccountOptions
+
+	// ClientMaxBalance defines the maximum balance a client may maintain before
+	// further Withdraw calls return zero.
+	ClientMaxBalance float64
+
+	// TTL defines the time to await further updates before purging entry for
+	// all accounts and clients.
+	TTL time.Duration
+	// PurgeInterval defines the amount of time to wait between purge calls.
+	PurgeInterval time.Duration
+}
+
+// Throttler manages the relationship between service instance clients and
+// backend storage credits. Each client must withdraw credits in order to submit
+// trace data to the Jaeger agent. These credits are generated on behalf of the
+// service, of which there can be many instances, and thus many clients.
+type Throttler struct {
+	sync.Mutex
+	accounts  map[string]*account
+	clients   map[string]*client
+	options   ThrottlerOptions
+	timeNow   func() time.Time // For testing
+	ticker    *time.Ticker
+	done      chan bool
+	waitGroup sync.WaitGroup
+}
+
+// NewThrottler creates a new throttler and returns it to the caller.
+// options should specify the specific throttler options to use.
+func NewThrottler(options ThrottlerOptions) *Throttler {
+	t := &Throttler{
+		accounts: map[string]*account{},
+		clients:  map[string]*client{},
+		options:  options,
+		timeNow:  time.Now,
+	}
+	t.ticker = time.NewTicker(t.options.PurgeInterval)
+	t.done = make(chan bool)
+	t.waitGroup.Add(1)
+	go func() {
+		defer t.waitGroup.Done()
+		for {
+			select {
+			case <-t.ticker.C:
+				t.purgeExpired()
+			case <-t.done:
+				return
+			}
+		}
+	}()
+	return t
+}
+
+// Withdraw deducts as many credits from the service on behalf of the client
+// without exceeding MaxClientBalance. The client is identified by a unique
+// clientID string. The credits are deducted from the specific operation on
+// the service if present, otherwise from the default credit pool.
+func (t *Throttler) Withdraw(serviceName string, clientID string, operationName string) float64 {
+	t.Lock()
+	defer t.Unlock()
+	c := t.findOrCreateClient(clientID)
+	a := t.findOrCreateAccount(serviceName)
+	balance := c.perOperationBalance[operationName]
+	maxWithdrawal := t.options.ClientMaxBalance - balance
```

Comment:
what are your plans for when the configs for a specific service are updated and the ClientMaxBalance < balance, maybe a max(0, ClientMaxBalance - balance) here for future proofing?

Reply:
I thought about it, but considered it an invalid case. Again, great place for an assertion if it existed. I can change it for now.